{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Differential (gene) Expression"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[intro...]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "differential_expression()\n",
    "\n",
    "Description\n",
    "- This module performs differential expression analysis between two or more phenotypes. For example, you could use this module to find the most differentially expressed genes between a groups of healthy and cancerous cell lines. \n",
    "Parameters\n",
    "- input_gene_expression\n",
    "  + Gene expression data filename (.gct or .txt file) or python DataFrame object where rows are genes and columns are samples.\n",
    "- phenotypes_labels\n",
    "  + Phenotype of each sample in input_gene_expression file (array, Series where indices match , or row of GCT file)\n",
    "- n_top_hits_to_display\n",
    "  + The number of most differentially expressed genes to show in the output plot (e.g., if this value is 5, a total of 10 genes (up/down-regulated) will be shown for each phenotype comparison).\n",
    "  + default = 20\n",
    "- ranking_method \n",
    "  + The function used to perform the differential expression analysis. The choice of ranking_method also specifies the statistical significance calculation.\n",
    "    - default = information_coefficient\n",
    "    - Pearson_correlation\n",
    "    - signal_to_noise_ratio\n",
    "    - t_test_statistic\n",
    "    - Spearman correlation\n",
    "- multi_phenotype_mode\n",
    "  + Perform multiple differential expression runs for multiple phenotypes and produces the same outputs as a 2 phenotype run but with results stacked (with “genes_to_show” x 2 entries for each comparison). When using multiple phenotypes the outputs (heatmap or tables) have the same format but show a stack of results corresponding to each class comparison.\n",
    "    - default = one_vs_all\n",
    "    - all_pairs\n",
    "- Outputs\n",
    "  + Differential_expression_heatmap\n",
    "    - A .png file of a heatmap showing the “n_top_hits_to_display” most differentially expressed genes in each phenotype comparison. Includes the confidence intervals for the metric and statistical significance results.\n",
    "  + Differential_expression_results\n",
    "    - A tab-separated TXT file which contains the results of the entire differential expression analysis, the confidence intervals for the metric and statistical significance results.\n",
    "- Returns\n",
    "  + This function returns (python) a data frame showing the same genes as the heatmap (gene_to_show x 2 entries).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Use these files:\n",
    "- http://datasets.genepattern.org/all_aml/all_aml_test.cls\n",
    "- http://datasets.genepattern.org/all_aml/all_aml_test.gct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "genepattern": {
     "output_variable": "de_results",
     "param_values": {
      "gene_expression": "http://datasets.genepattern.org/all_aml/all_aml_test.gct",
      "max_number_of_genes_to_show": "20",
      "number_of_permutations": "10",
      "output_filename": "Temp",
      "phenotypes": "http://datasets.genepattern.org/all_aml/all_aml_test.cls",
      "random_seed": "20121020",
      "ranking_method": "custom_pearson_corr",
      "title": "Temp_title"
     },
     "show_code": true,
     "type": "uibuilder"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4a007eaf7a944e2e8d0bedbab0342490",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<p>Failed to display Jupyter Widget of type <code>GPUIBuilder</code>.</p>\n",
       "<p>\n",
       "  If you're reading this message in the Jupyter Notebook or JupyterLab Notebook, it may mean\n",
       "  that the widgets JavaScript is still loading. If this message persists, it\n",
       "  likely means that the widgets JavaScript library is either not installed or\n",
       "  not enabled. See the <a href=\"https://ipywidgets.readthedocs.io/en/stable/user_install.html\">Jupyter\n",
       "  Widgets Documentation</a> for setup instructions.\n",
       "</p>\n",
       "<p>\n",
       "  If you're reading this message in another frontend (for example, a static\n",
       "  rendering on GitHub or <a href=\"https://nbviewer.jupyter.org/\">NBViewer</a>),\n",
       "  it may mean that your frontend doesn't currently support widgets.\n",
       "</p>\n"
      ],
      "text/plain": [
       "GPUIBuilder(description='Perform differential analysis on gene expression data of two phenotypes.', function_import='differential_gene_expression', name='Differential Gene Expression! =D', params=[{'name': 'phenotypes', 'label': 'phenotypes', 'optional': False, 'default': '', 'description': 'CLS filename; input binary phenotype/class distinction', 'hide': False, 'choices': []}, {'name': 'gene_expression', 'label': 'gene_expression', 'optional': False, 'default': '', 'description': 'GCT filename; data matrix with input gene expression profiles', 'hide': False, 'choices': []}, {'name': 'output_filename', 'label': 'output_filename', 'optional': False, 'default': '', 'description': 'Output files will have this name plus extensions .txt and .pdf', 'hide': False, 'choices': []}, {'name': 'ranking_method', 'label': 'ranking_method', 'optional': False, 'default': '', 'description': 'The function to use to compute similarity between phenotypes and gene_expression', 'hide': False, 'choices': []}, {'name': 'max_number_of_genes_to_show', 'label': 'max_number_of_genes_to_show', 'optional': True, 'default': 20, 'description': 'Maximum number of genes to show in the heatmap', 'hide': False, 'choices': []}, {'name': 'number_of_permutations', 'label': 'number_of_permutations', 'optional': True, 'default': 10, 'description': 'Number of random permutations to estimate statistical significance (p-values and FDRs)', 'hide': False, 'choices': []}, {'name': 'title', 'label': 'title', 'optional': True, 'default': None, 'description': 'The title of the heatmap', 'hide': False, 'choices': []}, {'name': 'random_seed', 'label': 'random_seed', 'optional': True, 'default': 20121020, 'description': 'Random number generator seed (can be set to a user supplied integer for reproducibility)', 'hide': False, 'choices': []}])"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import genepattern\n",
    "import cuzcatlan as cusca\n",
    "import pandas as pd\n",
    "import urllib.request\n",
    "from cuzcatlan import compute_information_coefficient\n",
    "from cuzcatlan import custom_pearson_corr\n",
    "import validators\n",
    "RANDOM_SEED = 20121020\n",
    "\n",
    "#Note: definining here for simplicity, but this is a mirror of the function in cuzcatlan.\n",
    "def differential_gene_expression(\n",
    "        phenotypes:\"CLS filename; input binary phenotype/class distinction\",\n",
    "        gene_expression:\"GCT filename; data matrix with input gene expression profiles\",\n",
    "        output_filename:\"Output files will have this name plus extensions .txt and .pdf\",\n",
    "        ranking_method:\"The function to use to compute similarity between phenotypes and gene_expression\",\n",
    "        max_number_of_genes_to_show:\"Maximum number of genes to show in the heatmap\"=20,\n",
    "        number_of_permutations:\"Number of random permutations to estimate statistical significance (p-values and FDRs)\"=10,\n",
    "        title:\"The title of the heatmap\"=None,\n",
    "        random_seed:\"Random number generator seed (can be set to a user supplied integer for reproducibility)\"=RANDOM_SEED):\n",
    "    \"\"\"\n",
    "    Perform differential analysis on gene expression data of two phenotypes.\n",
    "    \"\"\"\n",
    "    \n",
    "#     :param output_filename: \n",
    "#     :param ranking_method:callable;\n",
    "#     :param max_number_of_genes_to_show: int; \n",
    "#     :param number_of_permutations: int; \n",
    "#     :param title: str;\n",
    "#     :param random_seed: int | array; f\n",
    "#     :return: Dataframe; table of genes ranked by Information Coeff vs. phenotype\n",
    "\n",
    "    data_df = pd.read_table(gene_expression, header=2, index_col=0)\n",
    "    data_df.drop('Description', axis=1, inplace=True)\n",
    "    \n",
    "    if validators.url(phenotypes):\n",
    "        urlfile, __ = urllib.request.urlretrieve(phenotypes)\n",
    "    else:\n",
    "        urlfile = phenotypes\n",
    "        \n",
    "    temp = open(urlfile)\n",
    "    temp.readline()\n",
    "    temp.readline()\n",
    "    classes = [int(i) for i in temp.readline().strip('\\n').split(' ')]\n",
    "    classes = pd.Series(classes, index=data_df.columns)\n",
    "\n",
    "    gene_scores = cusca.make_match_panel(\n",
    "        features=data_df,\n",
    "        target=classes,\n",
    "        function=ranking_method,\n",
    "        target_ascending=False,\n",
    "        n_top_features=0.99,\n",
    "        max_n_features=max_number_of_genes_to_show,\n",
    "        n_samplings=30,\n",
    "        n_permutations=number_of_permutations,\n",
    "        random_seed=random_seed,\n",
    "        target_type='binary',\n",
    "        title=title,\n",
    "        file_path_prefix=output_filename)\n",
    "\n",
    "    return gene_scores\n",
    "\n",
    "# genepattern.GPUIBuilder(cusca.differential_gene_expression, name=\"Differential Gene Expression! =D\")\n",
    "genepattern.GPUIBuilder(differential_gene_expression, name=\"Differential Gene Expression! =D\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['ALL 19769 TA+ Norel', 'ALL 406 TA+ (ML) Norel', 'ALL 4466 Norel',\n",
      "       'ALL 1245 TA- Norel', 'ALL 16125 TA- Norel', 'ALL 23368 TA- Norel',\n",
      "       'ALL R28 (ML) Relap', 'ALL 1234 ML', 'ALL 1305 ML', 'ALL 1421 ML',\n",
      "       'ALL 1425 ML', 'ALL 1486 ML', 'ALL 1457 ML', 'ALL 1256 TA+ (ML) Norel',\n",
      "       'ALL 1275 TA+ (ML) Norel', 'ALL 1101 TA- (ML) Norel',\n",
      "       'ALL 1319 TA- (ML) Norel', 'ALL 522 TA- (ML) Norel', 'ALL SH 6',\n",
      "       'ALL SH 12', 'ALL SH 15', 'AML 15 (PK) Norel', 'AML 19 (PK) Norel',\n",
      "       'AML 10 (PK) Relap', 'AML 9 (PK) Relap', 'AML SH 5', 'AML SH 13',\n",
      "       'AML SH 14', 'AML SH 16', 'AML SH 18', 'AML 17 ML', 'AML 25 ML',\n",
      "       'AML 11 (ML)', 'AML 12 (ML)', 'AML 5 (ML) Relap'],\n",
      "      dtype='object')\n",
      "Index(['ALL 19769 TA+ Norel', 'ALL 406 TA+ (ML) Norel', 'ALL 4466 Norel',\n",
      "       'ALL 1245 TA- Norel', 'ALL 16125 TA- Norel', 'ALL 23368 TA- Norel',\n",
      "       'ALL R28 (ML) Relap', 'ALL 1234 ML', 'ALL 1305 ML', 'ALL 1421 ML',\n",
      "       'ALL 1425 ML', 'ALL 1486 ML', 'ALL 1457 ML', 'ALL 1256 TA+ (ML) Norel',\n",
      "       'ALL 1275 TA+ (ML) Norel', 'ALL 1101 TA- (ML) Norel',\n",
      "       'ALL 1319 TA- (ML) Norel', 'ALL 522 TA- (ML) Norel', 'ALL SH 6',\n",
      "       'ALL SH 12', 'ALL SH 15', 'AML 15 (PK) Norel', 'AML 19 (PK) Norel',\n",
      "       'AML 10 (PK) Relap', 'AML 9 (PK) Relap', 'AML SH 5', 'AML SH 13',\n",
      "       'AML SH 14', 'AML SH 16', 'AML SH 18', 'AML 17 ML', 'AML 25 ML',\n",
      "       'AML 11 (ML)', 'AML 12 (ML)', 'AML 5 (ML) Relap'],\n",
      "      dtype='object')\n",
      "AML 5 (ML) Relap           1\n",
      "AML SH 13                  1\n",
      "AML 12 (ML)                1\n",
      "AML 15 (PK) Norel          1\n",
      "AML 10 (PK) Relap          1\n",
      "AML 9 (PK) Relap           1\n",
      "AML SH 5                   1\n",
      "AML 19 (PK) Norel          1\n",
      "AML SH 14                  1\n",
      "AML SH 16                  1\n",
      "AML SH 18                  1\n",
      "AML 17 ML                  1\n",
      "AML 25 ML                  1\n",
      "AML 11 (ML)                1\n",
      "ALL 1425 ML                0\n",
      "ALL 406 TA+ (ML) Norel     0\n",
      "ALL 4466 Norel             0\n",
      "ALL 1245 TA- Norel         0\n",
      "ALL 16125 TA- Norel        0\n",
      "ALL 23368 TA- Norel        0\n",
      "ALL R28 (ML) Relap         0\n",
      "ALL 1234 ML                0\n",
      "ALL 1305 ML                0\n",
      "ALL 1421 ML                0\n",
      "ALL 522 TA- (ML) Norel     0\n",
      "ALL 1486 ML                0\n",
      "ALL 1457 ML                0\n",
      "ALL 1256 TA+ (ML) Norel    0\n",
      "ALL 1275 TA+ (ML) Norel    0\n",
      "ALL 1101 TA- (ML) Norel    0\n",
      "ALL 1319 TA- (ML) Norel    0\n",
      "ALL SH 6                   0\n",
      "ALL SH 12                  0\n",
      "ALL SH 15                  0\n",
      "ALL 19769 TA+ Norel        0\n",
      "dtype: int64\n",
      "Dropping 0 axis-1 slices ...\n",
      "[[-176. -137.  -48. ...,  -95.  -12. -214.]\n",
      " [-284.  -51. -531. ..., -118. -172. -153.]\n",
      " [ -81.  -82. -124. ...,   59.   12.  -58.]\n",
      " ..., \n",
      " [  20.  -11.    0. ...,  -43.  -68.   36.]\n",
      " [ 379.  198.  284. ...,  702.  226.  191.]\n",
      " [ -60.  -33.  -73. ...,   18.   78.  -37.]] [1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "Computing match score with <function custom_pearson_corr at 0x1108c0378> (1 process) ...\n",
      "Computing MoEs with 30 samplings ...\n",
      "Computing p-values and FDRs with 10 permutations ...\n",
      "\t1/10 ...\n",
      "\t2/10 ...\n",
      "\t3/10 ...\n",
      "\t4/10 ...\n",
      "\t5/10 ...\n",
      "\t6/10 ...\n",
      "\t7/10 ...\n",
      "\t8/10 ...\n",
      "\t9/10 ...\n",
      "\t10/10 ...\n",
      "\t10/10 - done.\n"
     ]
    }
   ],
   "source": [
    "de_results = differential_gene_expression(phenotypes=\"http://datasets.genepattern.org/all_aml/all_aml_test.cls\", gene_expression=\"http://datasets.genepattern.org/all_aml/all_aml_test.gct\", output_filename=\"Temp\", ranking_method=custom_pearson_corr, max_number_of_genes_to_show=20, number_of_permutations=10, title=\"Temp_title\", random_seed=20121020)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['ALL 19769 TA+ Norel', 'ALL 406 TA+ (ML) Norel', 'ALL 4466 Norel',\n",
      "       'ALL 1245 TA- Norel', 'ALL 16125 TA- Norel', 'ALL 23368 TA- Norel',\n",
      "       'ALL R28 (ML) Relap', 'ALL 1234 ML', 'ALL 1305 ML', 'ALL 1421 ML',\n",
      "       'ALL 1425 ML', 'ALL 1486 ML', 'ALL 1457 ML', 'ALL 1256 TA+ (ML) Norel',\n",
      "       'ALL 1275 TA+ (ML) Norel', 'ALL 1101 TA- (ML) Norel',\n",
      "       'ALL 1319 TA- (ML) Norel', 'ALL 522 TA- (ML) Norel', 'ALL SH 6',\n",
      "       'ALL SH 12', 'ALL SH 15', 'AML 15 (PK) Norel', 'AML 19 (PK) Norel',\n",
      "       'AML 10 (PK) Relap', 'AML 9 (PK) Relap', 'AML SH 5', 'AML SH 13',\n",
      "       'AML SH 14', 'AML SH 16', 'AML SH 18', 'AML 17 ML', 'AML 25 ML',\n",
      "       'AML 11 (ML)', 'AML 12 (ML)', 'AML 5 (ML) Relap'],\n",
      "      dtype='object')\n",
      "Index(['ALL 19769 TA+ Norel', 'ALL 406 TA+ (ML) Norel', 'ALL 4466 Norel',\n",
      "       'ALL 1245 TA- Norel', 'ALL 16125 TA- Norel', 'ALL 23368 TA- Norel',\n",
      "       'ALL R28 (ML) Relap', 'ALL 1234 ML', 'ALL 1305 ML', 'ALL 1421 ML',\n",
      "       'ALL 1425 ML', 'ALL 1486 ML', 'ALL 1457 ML', 'ALL 1256 TA+ (ML) Norel',\n",
      "       'ALL 1275 TA+ (ML) Norel', 'ALL 1101 TA- (ML) Norel',\n",
      "       'ALL 1319 TA- (ML) Norel', 'ALL 522 TA- (ML) Norel', 'ALL SH 6',\n",
      "       'ALL SH 12', 'ALL SH 15', 'AML 15 (PK) Norel', 'AML 19 (PK) Norel',\n",
      "       'AML 10 (PK) Relap', 'AML 9 (PK) Relap', 'AML SH 5', 'AML SH 13',\n",
      "       'AML SH 14', 'AML SH 16', 'AML SH 18', 'AML 17 ML', 'AML 25 ML',\n",
      "       'AML 11 (ML)', 'AML 12 (ML)', 'AML 5 (ML) Relap'],\n",
      "      dtype='object')\n",
      "AML 5 (ML) Relap           1\n",
      "AML SH 13                  1\n",
      "AML 12 (ML)                1\n",
      "AML 15 (PK) Norel          1\n",
      "AML 10 (PK) Relap          1\n",
      "AML 9 (PK) Relap           1\n",
      "AML SH 5                   1\n",
      "AML 19 (PK) Norel          1\n",
      "AML SH 14                  1\n",
      "AML SH 16                  1\n",
      "AML SH 18                  1\n",
      "AML 17 ML                  1\n",
      "AML 25 ML                  1\n",
      "AML 11 (ML)                1\n",
      "ALL 1425 ML                0\n",
      "ALL 406 TA+ (ML) Norel     0\n",
      "ALL 4466 Norel             0\n",
      "ALL 1245 TA- Norel         0\n",
      "ALL 16125 TA- Norel        0\n",
      "ALL 23368 TA- Norel        0\n",
      "ALL R28 (ML) Relap         0\n",
      "ALL 1234 ML                0\n",
      "ALL 1305 ML                0\n",
      "ALL 1421 ML                0\n",
      "ALL 522 TA- (ML) Norel     0\n",
      "ALL 1486 ML                0\n",
      "ALL 1457 ML                0\n",
      "ALL 1256 TA+ (ML) Norel    0\n",
      "ALL 1275 TA+ (ML) Norel    0\n",
      "ALL 1101 TA- (ML) Norel    0\n",
      "ALL 1319 TA- (ML) Norel    0\n",
      "ALL SH 6                   0\n",
      "ALL SH 12                  0\n",
      "ALL SH 15                  0\n",
      "ALL 19769 TA+ Norel        0\n",
      "dtype: int64\n",
      "Dropping 0 axis-1 slices ...\n",
      "[[-176. -137.  -48. ...,  -95.  -12. -214.]\n",
      " [-284.  -51. -531. ..., -118. -172. -153.]\n",
      " [ -81.  -82. -124. ...,   59.   12.  -58.]\n",
      " ..., \n",
      " [  20.  -11.    0. ...,  -43.  -68.   36.]\n",
      " [ 379.  198.  284. ...,  702.  226.  191.]\n",
      " [ -60.  -33.  -73. ...,   18.   78.  -37.]] [1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "Computing match score with <function custom_pearson_corr at 0x1139692f0> (1 process) ...\n",
      "Computing MoEs with 30 samplings ...\n",
      "Computing p-values and FDRs with 10 permutations ...\n",
      "\t1/10 ...\n",
      "\t2/10 ...\n",
      "\t3/10 ...\n",
      "\t4/10 ...\n",
      "\t5/10 ...\n",
      "\t6/10 ...\n",
      "\t7/10 ...\n",
      "\t8/10 ...\n",
      "\t9/10 ...\n",
      "\t10/10 ...\n",
      "\t10/10 - done.\n"
     ]
    }
   ],
   "source": [
    "de_results = differential_gene_expression(phenotypes=\"http://datasets.genepattern.org/all_aml/all_aml_test.cls\", gene_expression=\"http://datasets.genepattern.org/all_aml/all_aml_test.gct\", output_filename=\"Temp\", ranking_method=custom_pearson_corr, max_number_of_genes_to_show=20, number_of_permutations=10, title=\"Temp_title\", random_seed=20121020)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['ALL 19769 TA+ Norel', 'ALL 406 TA+ (ML) Norel', 'ALL 4466 Norel',\n",
      "       'ALL 1245 TA- Norel', 'ALL 16125 TA- Norel', 'ALL 23368 TA- Norel',\n",
      "       'ALL R28 (ML) Relap', 'ALL 1234 ML', 'ALL 1305 ML', 'ALL 1421 ML',\n",
      "       'ALL 1425 ML', 'ALL 1486 ML', 'ALL 1457 ML', 'ALL 1256 TA+ (ML) Norel',\n",
      "       'ALL 1275 TA+ (ML) Norel', 'ALL 1101 TA- (ML) Norel',\n",
      "       'ALL 1319 TA- (ML) Norel', 'ALL 522 TA- (ML) Norel', 'ALL SH 6',\n",
      "       'ALL SH 12', 'ALL SH 15', 'AML 15 (PK) Norel', 'AML 19 (PK) Norel',\n",
      "       'AML 10 (PK) Relap', 'AML 9 (PK) Relap', 'AML SH 5', 'AML SH 13',\n",
      "       'AML SH 14', 'AML SH 16', 'AML SH 18', 'AML 17 ML', 'AML 25 ML',\n",
      "       'AML 11 (ML)', 'AML 12 (ML)', 'AML 5 (ML) Relap'],\n",
      "      dtype='object')\n",
      "Index(['ALL 19769 TA+ Norel', 'ALL 406 TA+ (ML) Norel', 'ALL 4466 Norel',\n",
      "       'ALL 1245 TA- Norel', 'ALL 16125 TA- Norel', 'ALL 23368 TA- Norel',\n",
      "       'ALL R28 (ML) Relap', 'ALL 1234 ML', 'ALL 1305 ML', 'ALL 1421 ML',\n",
      "       'ALL 1425 ML', 'ALL 1486 ML', 'ALL 1457 ML', 'ALL 1256 TA+ (ML) Norel',\n",
      "       'ALL 1275 TA+ (ML) Norel', 'ALL 1101 TA- (ML) Norel',\n",
      "       'ALL 1319 TA- (ML) Norel', 'ALL 522 TA- (ML) Norel', 'ALL SH 6',\n",
      "       'ALL SH 12', 'ALL SH 15', 'AML 15 (PK) Norel', 'AML 19 (PK) Norel',\n",
      "       'AML 10 (PK) Relap', 'AML 9 (PK) Relap', 'AML SH 5', 'AML SH 13',\n",
      "       'AML SH 14', 'AML SH 16', 'AML SH 18', 'AML 17 ML', 'AML 25 ML',\n",
      "       'AML 11 (ML)', 'AML 12 (ML)', 'AML 5 (ML) Relap'],\n",
      "      dtype='object')\n",
      "AML 5 (ML) Relap           1\n",
      "AML SH 13                  1\n",
      "AML 12 (ML)                1\n",
      "AML 15 (PK) Norel          1\n",
      "AML 10 (PK) Relap          1\n",
      "AML 9 (PK) Relap           1\n",
      "AML SH 5                   1\n",
      "AML 19 (PK) Norel          1\n",
      "AML SH 14                  1\n",
      "AML SH 16                  1\n",
      "AML SH 18                  1\n",
      "AML 17 ML                  1\n",
      "AML 25 ML                  1\n",
      "AML 11 (ML)                1\n",
      "ALL 1425 ML                0\n",
      "ALL 406 TA+ (ML) Norel     0\n",
      "ALL 4466 Norel             0\n",
      "ALL 1245 TA- Norel         0\n",
      "ALL 16125 TA- Norel        0\n",
      "ALL 23368 TA- Norel        0\n",
      "ALL R28 (ML) Relap         0\n",
      "ALL 1234 ML                0\n",
      "ALL 1305 ML                0\n",
      "ALL 1421 ML                0\n",
      "ALL 522 TA- (ML) Norel     0\n",
      "ALL 1486 ML                0\n",
      "ALL 1457 ML                0\n",
      "ALL 1256 TA+ (ML) Norel    0\n",
      "ALL 1275 TA+ (ML) Norel    0\n",
      "ALL 1101 TA- (ML) Norel    0\n",
      "ALL 1319 TA- (ML) Norel    0\n",
      "ALL SH 6                   0\n",
      "ALL SH 12                  0\n",
      "ALL SH 15                  0\n",
      "ALL 19769 TA+ Norel        0\n",
      "dtype: int64\n",
      "Dropping 0 axis-1 slices ...\n",
      "[[-176. -137.  -48. ...,  -95.  -12. -214.]\n",
      " [-284.  -51. -531. ..., -118. -172. -153.]\n",
      " [ -81.  -82. -124. ...,   59.   12.  -58.]\n",
      " ..., \n",
      " [  20.  -11.    0. ...,  -43.  -68.   36.]\n",
      " [ 379.  198.  284. ...,  702.  226.  191.]\n",
      " [ -60.  -33.  -73. ...,   18.   78.  -37.]] [1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "Computing match score with <function custom_pearson_corr at 0x1110d7e18> (1 process) ...\n",
      "Computing MoEs with 30 samplings ...\n",
      "Computing p-values and FDRs with 10 permutations ...\n",
      "\t1/10 ...\n",
      "\t2/10 ...\n",
      "\t3/10 ...\n",
      "\t4/10 ...\n",
      "\t5/10 ...\n",
      "\t6/10 ...\n",
      "\t7/10 ...\n",
      "\t8/10 ...\n",
      "\t9/10 ...\n",
      "\t10/10 ...\n",
      "\t10/10 - done.\n"
     ]
    }
   ],
   "source": [
    "de_results = differential_gene_expression(phenotypes=\"http://datasets.genepattern.org/all_aml/all_aml_test.cls\", gene_expression=\"http://datasets.genepattern.org/all_aml/all_aml_test.gct\", output_filename=\"Temp\", ranking_method=custom_pearson_corr, max_number_of_genes_to_show=20, number_of_permutations=10, title=\"Temp_title\", random_seed=20121020)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['ALL 19769 TA+ Norel', 'ALL 406 TA+ (ML) Norel', 'ALL 4466 Norel',\n",
      "       'ALL 1245 TA- Norel', 'ALL 16125 TA- Norel', 'ALL 23368 TA- Norel',\n",
      "       'ALL R28 (ML) Relap', 'ALL 1234 ML', 'ALL 1305 ML', 'ALL 1421 ML',\n",
      "       'ALL 1425 ML', 'ALL 1486 ML', 'ALL 1457 ML', 'ALL 1256 TA+ (ML) Norel',\n",
      "       'ALL 1275 TA+ (ML) Norel', 'ALL 1101 TA- (ML) Norel',\n",
      "       'ALL 1319 TA- (ML) Norel', 'ALL 522 TA- (ML) Norel', 'ALL SH 6',\n",
      "       'ALL SH 12', 'ALL SH 15', 'AML 15 (PK) Norel', 'AML 19 (PK) Norel',\n",
      "       'AML 10 (PK) Relap', 'AML 9 (PK) Relap', 'AML SH 5', 'AML SH 13',\n",
      "       'AML SH 14', 'AML SH 16', 'AML SH 18', 'AML 17 ML', 'AML 25 ML',\n",
      "       'AML 11 (ML)', 'AML 12 (ML)', 'AML 5 (ML) Relap'],\n",
      "      dtype='object')\n",
      "Index(['ALL 19769 TA+ Norel', 'ALL 406 TA+ (ML) Norel', 'ALL 4466 Norel',\n",
      "       'ALL 1245 TA- Norel', 'ALL 16125 TA- Norel', 'ALL 23368 TA- Norel',\n",
      "       'ALL R28 (ML) Relap', 'ALL 1234 ML', 'ALL 1305 ML', 'ALL 1421 ML',\n",
      "       'ALL 1425 ML', 'ALL 1486 ML', 'ALL 1457 ML', 'ALL 1256 TA+ (ML) Norel',\n",
      "       'ALL 1275 TA+ (ML) Norel', 'ALL 1101 TA- (ML) Norel',\n",
      "       'ALL 1319 TA- (ML) Norel', 'ALL 522 TA- (ML) Norel', 'ALL SH 6',\n",
      "       'ALL SH 12', 'ALL SH 15', 'AML 15 (PK) Norel', 'AML 19 (PK) Norel',\n",
      "       'AML 10 (PK) Relap', 'AML 9 (PK) Relap', 'AML SH 5', 'AML SH 13',\n",
      "       'AML SH 14', 'AML SH 16', 'AML SH 18', 'AML 17 ML', 'AML 25 ML',\n",
      "       'AML 11 (ML)', 'AML 12 (ML)', 'AML 5 (ML) Relap'],\n",
      "      dtype='object')\n",
      "AML 5 (ML) Relap           1\n",
      "AML SH 13                  1\n",
      "AML 12 (ML)                1\n",
      "AML 15 (PK) Norel          1\n",
      "AML 10 (PK) Relap          1\n",
      "AML 9 (PK) Relap           1\n",
      "AML SH 5                   1\n",
      "AML 19 (PK) Norel          1\n",
      "AML SH 14                  1\n",
      "AML SH 16                  1\n",
      "AML SH 18                  1\n",
      "AML 17 ML                  1\n",
      "AML 25 ML                  1\n",
      "AML 11 (ML)                1\n",
      "ALL 1425 ML                0\n",
      "ALL 406 TA+ (ML) Norel     0\n",
      "ALL 4466 Norel             0\n",
      "ALL 1245 TA- Norel         0\n",
      "ALL 16125 TA- Norel        0\n",
      "ALL 23368 TA- Norel        0\n",
      "ALL R28 (ML) Relap         0\n",
      "ALL 1234 ML                0\n",
      "ALL 1305 ML                0\n",
      "ALL 1421 ML                0\n",
      "ALL 522 TA- (ML) Norel     0\n",
      "ALL 1486 ML                0\n",
      "ALL 1457 ML                0\n",
      "ALL 1256 TA+ (ML) Norel    0\n",
      "ALL 1275 TA+ (ML) Norel    0\n",
      "ALL 1101 TA- (ML) Norel    0\n",
      "ALL 1319 TA- (ML) Norel    0\n",
      "ALL SH 6                   0\n",
      "ALL SH 12                  0\n",
      "ALL SH 15                  0\n",
      "ALL 19769 TA+ Norel        0\n",
      "dtype: int64\n",
      "Dropping 0 axis-1 slices ...\n",
      "[[-176. -137.  -48. ...,  -95.  -12. -214.]\n",
      " [-284.  -51. -531. ..., -118. -172. -153.]\n",
      " [ -81.  -82. -124. ...,   59.   12.  -58.]\n",
      " ..., \n",
      " [  20.  -11.    0. ...,  -43.  -68.   36.]\n",
      " [ 379.  198.  284. ...,  702.  226.  191.]\n",
      " [ -60.  -33.  -73. ...,   18.   78.  -37.]] [1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "Computing match score with <function custom_pearson_corr at 0x109e94d90> (1 process) ...\n",
      "Computing MoEs with 30 samplings ...\n",
      "Computing p-values and FDRs with 10 permutations ...\n",
      "\t1/10 ...\n",
      "\t2/10 ...\n",
      "\t3/10 ...\n",
      "\t4/10 ...\n",
      "\t5/10 ...\n",
      "\t6/10 ...\n",
      "\t7/10 ...\n",
      "\t8/10 ...\n",
      "\t9/10 ...\n",
      "\t10/10 ...\n",
      "\t10/10 - done.\n"
     ]
    }
   ],
   "source": [
    "de_results = differential_gene_expression(phenotypes=\"http://datasets.genepattern.org/all_aml/all_aml_test.cls\", gene_expression=\"http://datasets.genepattern.org/all_aml/all_aml_test.gct\", output_filename=\"Temp\", ranking_method=custom_pearson_corr, max_number_of_genes_to_show=20, number_of_permutations=10, title=\"Temp_title\", random_seed=20121020)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Score</th>\n",
       "      <th>0.95 MoE</th>\n",
       "      <th>p-value</th>\n",
       "      <th>FDR</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>U46499_at</th>\n",
       "      <td>0.824544</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X95735_at</th>\n",
       "      <td>0.775671</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M63959_at</th>\n",
       "      <td>0.769661</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L09209_s_at</th>\n",
       "      <td>0.747575</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M84526_at</th>\n",
       "      <td>0.740053</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X17042_at</th>\n",
       "      <td>0.731650</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M14636_at</th>\n",
       "      <td>0.726834</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M22960_at</th>\n",
       "      <td>0.726074</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U59878_at</th>\n",
       "      <td>0.720990</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M23197_at</th>\n",
       "      <td>0.716258</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X61587_at</th>\n",
       "      <td>0.711473</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M27891_at</th>\n",
       "      <td>0.706726</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X05908_at</th>\n",
       "      <td>0.704679</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U60319_at</th>\n",
       "      <td>0.700721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M32304_s_at</th>\n",
       "      <td>0.690860</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X12447_at</th>\n",
       "      <td>0.688066</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M19507_at</th>\n",
       "      <td>0.686567</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X16546_at</th>\n",
       "      <td>0.684451</td>\n",
       "      <td>0.0291901</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M63138_at</th>\n",
       "      <td>0.682516</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L09717_at</th>\n",
       "      <td>0.680947</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Z29067_at</th>\n",
       "      <td>0.673727</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L41559_at</th>\n",
       "      <td>0.672050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>J02783_at</th>\n",
       "      <td>0.662745</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U05572_s_at</th>\n",
       "      <td>0.654224</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X52056_at</th>\n",
       "      <td>0.653856</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M16038_at</th>\n",
       "      <td>0.649158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D11327_s_at</th>\n",
       "      <td>0.643376</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M62762_at</th>\n",
       "      <td>0.639631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U10868_at</th>\n",
       "      <td>0.637952</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X62654_rna1_at</th>\n",
       "      <td>0.635277</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X98833_rna1_at</th>\n",
       "      <td>-0.592751</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X97267_rna1_s_at</th>\n",
       "      <td>-0.593663</td>\n",
       "      <td>0.0261791</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X67951_at</th>\n",
       "      <td>-0.595051</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M28170_at</th>\n",
       "      <td>-0.597141</td>\n",
       "      <td>0.0345944</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>J02874_at</th>\n",
       "      <td>-0.597197</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U79277_at</th>\n",
       "      <td>-0.597840</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X69111_at</th>\n",
       "      <td>-0.599072</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M65214_s_at</th>\n",
       "      <td>-0.605832</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M98045_at</th>\n",
       "      <td>-0.617877</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U88964_at</th>\n",
       "      <td>-0.620409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U07139_at</th>\n",
       "      <td>-0.625601</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y08612_at</th>\n",
       "      <td>-0.625998</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HG1612-HT1612_at</th>\n",
       "      <td>-0.628046</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X58529_at</th>\n",
       "      <td>-0.630023</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X59417_at</th>\n",
       "      <td>-0.633434</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Z49194_at</th>\n",
       "      <td>-0.633840</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X66401_cds1_at</th>\n",
       "      <td>-0.640256</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M33680_at</th>\n",
       "      <td>-0.642720</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U51240_at</th>\n",
       "      <td>-0.645048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D30742_at</th>\n",
       "      <td>-0.645910</td>\n",
       "      <td>0.0257204</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M31211_s_at</th>\n",
       "      <td>-0.667406</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D88270_at</th>\n",
       "      <td>-0.674907</td>\n",
       "      <td>0.0228069</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M92287_at</th>\n",
       "      <td>-0.675219</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U05259_rna1_at</th>\n",
       "      <td>-0.675521</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U29175_at</th>\n",
       "      <td>-0.680743</td>\n",
       "      <td>0.0185103</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M11722_at</th>\n",
       "      <td>-0.718129</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M31523_at</th>\n",
       "      <td>-0.722196</td>\n",
       "      <td>0.0228747</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M84371_rna1_s_at</th>\n",
       "      <td>-0.729815</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>J05243_at</th>\n",
       "      <td>-0.737162</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M89957_at</th>\n",
       "      <td>-0.775614</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.002273</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7129 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Score   0.95 MoE   p-value       FDR\n",
       "Name                                                     \n",
       "U46499_at         0.824544        NaN  0.000014  0.002381\n",
       "X95735_at         0.775671        NaN  0.000014  0.002381\n",
       "M63959_at         0.769661        NaN  0.000014  0.002381\n",
       "L09209_s_at       0.747575        NaN  0.000014  0.002381\n",
       "M84526_at         0.740053        NaN  0.000014  0.002381\n",
       "X17042_at         0.731650        NaN  0.000014  0.002381\n",
       "M14636_at         0.726834        NaN  0.000014  0.002381\n",
       "M22960_at         0.726074        NaN  0.000014  0.002381\n",
       "U59878_at         0.720990        NaN  0.000014  0.002381\n",
       "M23197_at         0.716258        NaN  0.000014  0.002381\n",
       "X61587_at         0.711473        NaN  0.000014  0.002381\n",
       "M27891_at         0.706726        NaN  0.000014  0.002381\n",
       "X05908_at         0.704679        NaN  0.000014  0.002381\n",
       "U60319_at         0.700721        NaN  0.000014  0.002381\n",
       "M32304_s_at       0.690860        NaN  0.000014  0.002381\n",
       "X12447_at         0.688066        NaN  0.000014  0.002381\n",
       "M19507_at         0.686567        NaN  0.000014  0.002381\n",
       "X16546_at         0.684451  0.0291901  0.000014  0.002381\n",
       "M63138_at         0.682516        NaN  0.000014  0.002381\n",
       "L09717_at         0.680947        NaN  0.000014  0.002381\n",
       "Z29067_at         0.673727        NaN  0.000014  0.002381\n",
       "L41559_at         0.672050        NaN  0.000014  0.002381\n",
       "J02783_at         0.662745        NaN  0.000014  0.002381\n",
       "U05572_s_at       0.654224        NaN  0.000014  0.002381\n",
       "X52056_at         0.653856        NaN  0.000014  0.002381\n",
       "M16038_at         0.649158        NaN  0.000014  0.002381\n",
       "D11327_s_at       0.643376        NaN  0.000014  0.002381\n",
       "M62762_at         0.639631        NaN  0.000014  0.002381\n",
       "U10868_at         0.637952        NaN  0.000014  0.002381\n",
       "X62654_rna1_at    0.635277        NaN  0.000014  0.002381\n",
       "...                    ...        ...       ...       ...\n",
       "X98833_rna1_at   -0.592751        NaN  0.000014  0.002273\n",
       "X97267_rna1_s_at -0.593663  0.0261791  0.000014  0.002273\n",
       "X67951_at        -0.595051        NaN  0.000014  0.002273\n",
       "M28170_at        -0.597141  0.0345944  0.000014  0.002273\n",
       "J02874_at        -0.597197        NaN  0.000014  0.002273\n",
       "U79277_at        -0.597840        NaN  0.000014  0.002273\n",
       "X69111_at        -0.599072        NaN  0.000014  0.002273\n",
       "M65214_s_at      -0.605832        NaN  0.000014  0.002273\n",
       "M98045_at        -0.617877        NaN  0.000014  0.002273\n",
       "U88964_at        -0.620409        NaN  0.000014  0.002273\n",
       "U07139_at        -0.625601        NaN  0.000014  0.002273\n",
       "Y08612_at        -0.625998        NaN  0.000014  0.002273\n",
       "HG1612-HT1612_at -0.628046        NaN  0.000014  0.002273\n",
       "X58529_at        -0.630023        NaN  0.000014  0.002273\n",
       "X59417_at        -0.633434        NaN  0.000014  0.002273\n",
       "Z49194_at        -0.633840        NaN  0.000014  0.002273\n",
       "X66401_cds1_at   -0.640256        NaN  0.000014  0.002273\n",
       "M33680_at        -0.642720        NaN  0.000014  0.002273\n",
       "U51240_at        -0.645048        NaN  0.000014  0.002273\n",
       "D30742_at        -0.645910  0.0257204  0.000014  0.002273\n",
       "M31211_s_at      -0.667406        NaN  0.000014  0.002273\n",
       "D88270_at        -0.674907  0.0228069  0.000014  0.002273\n",
       "M92287_at        -0.675219        NaN  0.000014  0.002273\n",
       "U05259_rna1_at   -0.675521        NaN  0.000014  0.002273\n",
       "U29175_at        -0.680743  0.0185103  0.000014  0.002273\n",
       "M11722_at        -0.718129        NaN  0.000014  0.002273\n",
       "M31523_at        -0.722196  0.0228747  0.000014  0.002273\n",
       "M84371_rna1_s_at -0.729815        NaN  0.000014  0.002273\n",
       "J05243_at        -0.737162        NaN  0.000014  0.002273\n",
       "M89957_at        -0.775614        NaN  0.000014  0.002273\n",
       "\n",
       "[7129 rows x 4 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "de_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
